From 3289415875def51ccc6ffe0334a3818920e58beb Mon Sep 17 00:00:00 2001
From: Jacopo Mondi <jacopo+renesas@jmondi.org>
Date: Tue, 29 May 2018 10:48:02 +0200
Subject: [PATCH 354/909] media: rcar-vin: Cleanup notifier in error path

commit 8438bfa6281d7f8f9775a085a54196efa27b04e3 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

During the notifier initialization, memory for the list of associated async
subdevices is reserved during the fwnode endpoint parsing from the
v4l2-async framework. If the notifier registration fails, that memory
should be released and the notifier 'cleaned up'.

Catch the notifier registration error and perform the cleanup both for the
group and the parallel notifiers.

Link: https://patchwork.kernel.org/patch/10434533/
[koji.matsuoka.xm: the part of the patch for kernel v4.14.70 was rebased
and imported]
Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/media/platform/rcar-vin/rcar-core.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/drivers/media/platform/rcar-vin/rcar-core.c b/drivers/media/platform/rcar-vin/rcar-core.c
index b1952071f855..3d38158876c1 100644
--- a/drivers/media/platform/rcar-vin/rcar-core.c
+++ b/drivers/media/platform/rcar-vin/rcar-core.c
@@ -577,6 +577,7 @@ static int rvin_digital_graph_init(struct rvin_dev *vin)
 	ret = v4l2_async_notifier_register(&vin->v4l2_dev, &vin->notifier);
 	if (ret < 0) {
 		vin_err(vin, "Notifier registration failed\n");
+		v4l2_async_notifier_cleanup(&vin->notifier);
 		return ret;
 	}
 
@@ -799,6 +800,7 @@ static int rvin_mc_parse_of_graph(struct rvin_dev *vin)
 	ret = v4l2_async_notifier_register(&vin->v4l2_dev, &vin->notifier);
 	if (ret < 0) {
 		vin_err(vin, "Notifier registration failed\n");
+		v4l2_async_notifier_cleanup(&vin->notifier);
 		return ret;
 	}
 
-- 
2.17.1

